### PR TITLE
Warming up

### PR DESCRIPTION
**Pull request by jasmingacic**
@hasheddan I'm just warming up. 
So I added some fields to the update function. Also I had to update the Packet SDK.

You will see that I've been adding connection information into `DeviceStatus` instead of `DeviceSpec` but it shouldn't be a huge deal for the time being.

### CODE REVIEW

File: Gopkg.lock
Comment (hasheddan) [inline Gopkg.lock]: Curious as to why the revision is blank here. To move this to v0.2.0 locally I ran `dep ensure -update github.com/packethost/packngo` and it updated revision for me as well (`revision = "02a3ebac54fcfe6297c72ddfb68f8144f31c47b4"`). Would you mind giving that a try?

File: api/server/v1alpha1/device_types.go
Comment (hasheddan) [inline api/server/v1alpha1/device_types.go]: In looking through `packngo` it appears that `Locked` is present in [`DeviceUpdateRequest`](https://github.com/packethost/packngo/blob/6292a62434aeffca0d2cca5c2079349c2fc64650/devices.go#L166) but not `DeviceCreateRequest`. From what I can tell, you cannot create a device that is locked, but you can lock a created device. So if a user specifies `locked: true` in their spec on create here, then it would just get locked on the first call to `Update()` after the device is provisioned. Is my logic correct here?

File: pkg/clients/device/device.go
Comment (hasheddan) [inline pkg/clients/device/device.go:40]: Should we change this to `[]string` in spec?

File: pkg/clients/packet.go
Comment (hasheddan) [inline pkg/clients/packet.go:28]: Was this a change in `packngo` v0.2.0?

File: pkg/controller/packet/server/device/managed.go
Comment (hasheddan) [inline pkg/controller/packet/server/device/managed.go]: Since we do this in the `Observe()` we don't need to do so here. This `TODO` comment can be removed though! Not sure why I put it here as I was actually referring to connection details to connect to the instance, which should be propagated to a Kubernetes `Secret` in `Observe()` (see example [here](https://github.com/crossplaneio/stack-gcp/blob/48dc96d9685094e9a1fc3f2eab321991fa46f33d/pkg/controller/cache/managed.go#L135)). Not sure exactly what we would want to put into the secret for a device, SSH info, ports, IP?

File: pkg/controller/packet/server/device/managed.go
Comment (hasheddan) [inline pkg/controller/packet/server/device/managed.go]: I usually like to have a func in `clients/` that is something like `device.NeedsUpdate()` (example [here](https://github.com/crossplaneio/stack-gcp/blob/48dc96d9685094e9a1fc3f2eab321991fa46f33d/pkg/clients/cloudmemorystore/cloudmemorystore.go#L136)) just to keep the controller a little cleaner. Totally a personal preference though!

**Review by hasheddan — COMMENTED** [review]
This looks good! A few small comments to address but then will feel good about merging :) Nice work!

File: Gopkg.lock
Comment (jasmingacic) [reply Gopkg.lock]: I just ran that and it seems it has updated the revision.

File: api/server/v1alpha1/device_types.go
Comment (jasmingacic) [reply api/server/v1alpha1/device_types.go]: Yes you are correct. Create request does not accept `Locked` parameter.

File: pkg/clients/device/device.go
Comment (jasmingacic) [reply pkg/clients/device/device.go:40]: We could do that, but I'm not sure how practical it will be.

File: pkg/clients/packet.go
Comment (jasmingacic) [reply pkg/clients/packet.go:28]: Not just that one. UpdateDeviceRequest was not mapped correctly before. 
```
// DeviceUpdateRequest type used to update a Packet device
type DeviceUpdateRequest struct {
	Hostname      string   `json:"hostname,omitempty"`
```
vs
```
// DeviceUpdateRequest type used to update a Packet device
type DeviceUpdateRequest struct {
	Hostname      *string   `json:"hostname,omitempty"`
```

File: pkg/controller/packet/server/device/managed.go
Comment (jasmingacic) [reply pkg/controller/packet/server/device/managed.go]: I would go for now with public addresses IPv4 and IPv6, we can also surface the public ssh_key but there is a catch. When creating a device the device get injected with several SSH keys, if configured, so we could end up with several public SSH keys. 

I'd go with IPv4 only. What do you think?

File: pkg/controller/packet/server/device/managed.go
Comment (jasmingacic) [reply pkg/controller/packet/server/device/managed.go]: I agree it should be in a different function. I will encapsulate this so it is neater. I just wanted to get this started.

File: pkg/clients/device/device.go
Comment (hasheddan) [reply pkg/clients/device/device.go:40]: Can you specify multiple facilities in a create request? Does that provision multiple devices?

Comment (jasmingacic) [reply pkg/clients/device/device.go:40]: You can specify multiple facilities but that only creates a single device in one of the provided facilities. For example { "facility": ["f3", "f2", "any"] } will try to assign to the facility f3, if there are no available f2, and so on. If "any" is not specified for "facility", the request will fail unless it can assign in the selected locations.

File: pkg/controller/packet/server/device/managed.go
Comment (hasheddan) [reply pkg/controller/packet/server/device/managed.go]: hmm if we are going to reflect IP addresses in the status, maybe we don't need to propagate a secret?

Comment (jasmingacic) [reply pkg/controller/packet/server/device/managed.go]: In this case I don't see any use in propagating public SSH keys. I'd stick with just IPv4 and nothing else.

Comment (jasmingacic) [reply pkg/controller/packet/server/device/managed.go]: I've left only IPv4.

File: pkg/clients/device/device.go
Comment (hasheddan) [reply pkg/clients/device/device.go:40]: Gotcha! I am fine with this behavior for now, but I would recommend that we reflect the exact possible configuration in the future, as the current implementation does expose the full functionality of the Packet API.

File: pkg/clients/device/device.go
Comment (hasheddan) [inline pkg/clients/device/device.go:51]: Hmm I wasn't aware that other fields would cause delete and recreate. I agree that we do not want to do that in normal reconciliation. However, I think we would want to reflect somehow that the external resource has drifted from the desired spec but is not going to be changed by the controller.

File: pkg/controller/packet/packet.go
Comment (hasheddan) [inline pkg/controller/packet/packet.go:30]: This naming is a bit redundant (`device.DeviceController`) as I am not sure there will be another controller in the device package, but I am totally fine with it if you feel it is best :)

File: pkg/clients/device/device.go
Comment (hasheddan) [reply pkg/clients/device/device.go:51]: We are looking at addressing this issue here crossplaneio/crossplane-runtime#39

File: api/server/v1alpha1/device_types.go
Comment (hasheddan) [inline api/server/v1alpha1/device_types.go:163]: The `DeviceClass` will need required methods like those generated here (https://github.com/crossplaneio/stack-gcp/blob/master/apis/cache/v1alpha2/zz_generated.nonportableclass.go) in order to satisfy the `NonPortableClass` interface. We have a library to do this now and I would be happy to make the required updates after this PR is merged if that is easier.

File: api/server/v1alpha1/device_types.go
Comment (hasheddan) [inline api/server/v1alpha1/device_types.go]: Because `DeviceSpec` has `runtimev1alpha1.ResourceSpec` embedded in it and the `DeviceClass` is not a managed resource (it is a non-portable class), we usually separate out the fields of the provider-specific spec fields into a different struct called something like `DeviceParameters` and embed that in the `DeviceSpec` and `DeviceClassSpecTemplate`.

Ref: https://github.com/crossplaneio/stack-gcp/blob/32c2999e25af12712aa811658c4526d084838944/apis/cache/v1alpha2/cloudmemorystore_instance_types.go#L109

Ref: https://github.com/crossplaneio/stack-gcp/blob/32c2999e25af12712aa811658c4526d084838944/apis/cache/v1alpha2/cloudmemorystore_instance_types.go#L172

Ref: https://github.com/crossplaneio/stack-gcp/blob/32c2999e25af12712aa811658c4526d084838944/apis/cache/v1alpha2/cloudmemorystore_instance_types.go#L46

**Review by hasheddan — COMMENTED** [review]
Looking really good! There are some comments to address, but I would be happy to fix some of them in a follow-up PR if that would be easier :) feel free to weigh-in then we can talk about if we want to go ahead and merge!

File: api/server/v1alpha1/device_types.go
Comment (jasmingacic) [reply api/server/v1alpha1/device_types.go:163]: If you could that I'd be grateful. Or if you can point me in the direction how to do it.

File: api/server/v1alpha1/device_types.go
Comment (jasmingacic) [reply api/server/v1alpha1/device_types.go]: I've added requested changes essentially I encapsulated `DeviceParameters` and used it in `DeviceClassTemplate` and `DeviceSpec`.

File: pkg/clients/device/device.go
Comment (jasmingacic) [reply pkg/clients/device/device.go:51]: I think that part of that comment I pasted when I used other providers as a template.

File: pkg/controller/packet/packet.go
Comment (jasmingacic) [reply pkg/controller/packet/packet.go:30]: We could go with `PacketDeviceController` if you think it is better.

File: api/server/v1alpha1/device_types.go
Comment (hasheddan) [reply api/server/v1alpha1/device_types.go:163]: Sure thing! Would be happy to

File: pkg/clients/device/device.go
Comment (hasheddan) [reply pkg/clients/device/device.go:51]: Oh gotcha! Do you think that we should check every field that can be updated as part of [`DeviceUpdateRequest`](https://github.com/packethost/packngo/blob/6292a62434aeffca0d2cca5c2079349c2fc64650/devices.go#L166) then?

File: pkg/clients/device/device.go
Comment (hasheddan) [inline pkg/clients/device/device.go:78]: Similar to comment above, should we include everything in [`DeviceUpdateRequest`](https://github.com/packethost/packngo/blob/6292a62434aeffca0d2cca5c2079349c2fc64650/devices.go#L166)?

File: pkg/controller/packet/packet.go
Comment (hasheddan) [reply pkg/controller/packet/packet.go:30]: I was more suggesting just keeping `device.Controller`, but I prefer `device.DeviceController` over `device.PacketDeviceController`.

File: pkg/clients/device/device.go
Comment (jasmingacic) [reply pkg/clients/device/device.go:78]: As I extended `DeviceProperties` now we can do the check. Although I think if only one is changed then we can proceed to update all. The operation would be idempotent.

Comment (hasheddan) [reply pkg/clients/device/device.go:78]: Right, I agree we can update all, but here we are only choosing to update `Hostname` and `Locked`, while our `spec` includes many other fields which can also be included in a `DeviceUpdateRequest`.

Comment (jasmingacic) [reply pkg/clients/device/device.go:78]: I'll let you know once I have it updated.

**Review by hasheddan — APPROVED** [review]

File: api/server/v1alpha1/device_types.go
Comment (jasmingacic) [reply api/server/v1alpha1/device_types.go:163]: @hasheddan Do you need any help with the updates you mentioned above?

Comment (hasheddan) [reply api/server/v1alpha1/device_types.go:163]: @jasmingacic I believe I implemented these updates in #2. We have a few changes coming this week that will also require updates that I would be happy to implement :) Also, do you know if it is going to be possible to get this repo name changed back to `stack-packet`?

**Pull request merged**
